mods/dwm-ghost-mods.wh.cpp

```diff
@@ -0,0 +1,499 @@
+{
+    // If DWM frames are used, then the position of the thumbnail needs to be moved or it will
+    // render at { 0, 0 } inside the window frames.
+    if (g_prefDisableDwmFrames)
+    {
+        SIZE thumbSrcSize;
+        DwmQueryThumbnailSourceSize(hThumbnail, &thumbSrcSize);
+
+        NCWindowSizes ncWindowSizes;
+        GetNCWindowSizes(&ncWindowSizes);
+
+        pProperties->dwFlags |= DWM_TNP_RECTDESTINATION | DWM_TNP_RECTSOURCE;
+
+        // Calculate source positions
+        pProperties->rcSource.top = ncWindowSizes.topBorderHeight;
+        pProperties->rcSource.left = ncWindowSizes.leftRightBottomBorderSize;
+        pProperties->rcSource.right = thumbSrcSize.cx - ncWindowSizes.leftRightBottomBorderSize;
+        pProperties->rcSource.bottom = thumbSrcSize.cy - ncWindowSizes.leftRightBottomBorderSize;
+
+        // We reuse the same values that we used to crop the source to translate the destination.
+        // No additional work is required here.
+        pProperties->rcDestination = pProperties->rcSource;
+    }
+
+    if (g_prefFrostPolicy == FrostPolicyType::BASIC)
+    {
+        if (pProperties->opacity < 255)
+        {
+            pProperties->opacity = 128;
+        }
+    }
+    else if (g_prefFrostPolicy == FrostPolicyType::DISABLED)
+    {
+        if (pProperties->opacity < 255)
+        {
+            pProperties->opacity = 255;
+        }
+    }
+
+    return DwmUpdateThumbnailProperties_orig(hThumbnail, pProperties);
+}
+
+void LoadSettings()
+{
+    g_prefDisableDwmFrames = Wh_GetIntSetting(L"disableDwmFrames");
+    g_prefUseClassicTheme = Wh_GetIntSetting(L"useClassicTheme");
+
+    PCWSTR pszFrostPolicy = Wh_GetStringSetting(L"frostPolicy");
+    
+    if (lstrcmpW(pszFrostPolicy, L"enabled") == 0)
+    {
+        g_prefFrostPolicy = FrostPolicyType::ENABLED;
+    }
+    else if (lstrcmpW(pszFrostPolicy, L"basic") == 0)
+    {
+        g_prefFrostPolicy = FrostPolicyType::BASIC;
+    }
+    else
+    {
+        g_prefFrostPolicy = FrostPolicyType::DISABLED;
+    }
+
+    Wh_FreeStringSetting(pszFrostPolicy);
+
+    // Validate:
+    if (g_prefUseClassicTheme && !g_prefDisableDwmFrames)
+    {
+        DisplayErrorMessage(
+            L"Loaded invalid settings. In order to enable the classic theme on dialogs, "
+            L"you must also enable the \"Disable DWM frames\" setting. Please enable that "
+            L"option and try again."
+        );
+
+        g_prefUseClassicTheme = FALSE;
+    }
+}
+
+// The mod is being initialized, load settings, hook functions, and do other
+// initialization stuff if required.
+BOOL Wh_ModInit()
+{
+    #ifdef _WIN64
+    const int offset_same_teb_flags = 0x17EE;
+    #define NT_CURRENT_TEB_T __int64
+    #else
+    const int offset_same_teb_flags = 0x0FCA;
+    #define NT_CURRENT_TEB_T long
+    #endif
+
+    // 0x0400 = InitialThread
+    bool already_running = (!*(USHORT *)(((NT_CURRENT_TEB_T *)NtCurrentTeb()) + offset_same_teb_flags) & 0x0400);
```

Comment:
The compiler shows a warning on this line:

> Logical not is only applied to the left hand side of this bitwise operator (fixes available)

As it's written, it's likely to always return false:

```cpp
(!*(USHORT *)(((NT_CURRENT_TEB_T *)NtCurrentTeb()) + offset_same_teb_flags) & 0x0400);
(!*(USHORT *)((x1) + offset_same_teb_flags) & 0x0400);
(!*(USHORT *)(x2) & 0x0400);
(!*(USHORT *)(x3) & 0x0400);
(!x4 & 0x0400);
(1 & 0x0400);
0;
```

Is it even necessary outside of testing?

Reply:
Removed. I added it as a safeguard when I started development and never removed it; didn't even notice the typo.